src/core/context/context-management/ContextManager.ts

```diff
@@ -788,21 +826,28 @@ export class ContextManager {
 			const message = apiMessages[i]
 			if (message.role === "user" && Array.isArray(message.content) && message.content.length > 0) {
 				const firstBlock = message.content[0]
-				if (firstBlock.type === "text") {
-					const result = this.parseToolCallWithFormat(firstBlock.text)
+				// Extract text from either a direct text block or from inside a tool_result wrapper (native tool calling)
+				const firstBlockText = this.getTextFromBlock(firstBlock)
+
+				if (firstBlockText) {
+					const result = this.parseToolCallWithFormat(firstBlockText)
 					let foundNormalFileRead = false
 					if (result) {
 						const [toolName, filePath, contentBlockIndex, headerText] = result
 
 						if (toolName === "read_file") {
+							// For native tool calling format, we assume contentBlockIndex=0 which is what happens naturally
 							this.handleReadFileToolCall(i, filePath, fileReadIndices, contentBlockIndex, headerText)
 							foundNormalFileRead = true
 						} else if (toolName === "replace_in_file" || toolName === "write_to_file") {
-							// old format has the file contents in index=1 whereas the new format has it in index=0
-							// in either case we need to extract the correct contents
+							// For native tool calling format, the content is assumed to always in the same block (index=0 inside tool_result)
+							// For the XML format, the old format has the file contents in index=1 whereas the new format has it in index=0
 							let blockText: string | undefined
-							if (contentBlockIndex == 0) {
-								blockText = firstBlock.text
+							if (firstBlock.type === "tool_result") {
+								blockText = firstBlockText
+							} else if (contentBlockIndex == 0) {
```

Comment:
lint: `Using == may be unsafe if you are relying on type coercion.`
```suggestion
							} else if (contentBlockIndex === 0) {
```

Reply:
<img width="1475" height="267" alt="image" src="https://github.com/user-attachments/assets/3285fba0-3ecb-420b-8f89-313fc44fcb3a" />